### DISCUSSION TITLE
Text wrapping on buttons and menus

### DISCUSSION THREAD

**Discussion by treeder** [material-components/material-web#5176] [General]
When a button or menu item has text with a space in it, it will default wrap like this:

![image](https://github.com/material-components/material-web/assets/75826/fdefab71-1a75-4351-9920-7c2b24756031)

![image](https://github.com/material-components/material-web/assets/75826/e3189c78-e8f4-449d-8844-d31fd7d392e2)

If something else is a bit wider to make the menu wider, it won't. 

![image](https://github.com/material-components/material-web/assets/75826/e39e0e8e-c1a6-41f0-820c-c985f330b26b)

Of I can add `&nbsp;` instead of space:  

![image](https://github.com/material-components/material-web/assets/75826/062cce0e-b9c3-4541-906c-94fd9adcb7ef)

Is this expected behavior? I'd kind of expect it to never wrap (unless I explicitly wanted it to).

**Comment by asyncliz** [comment]
[This](https://github.com/material-components/material-web/pull/5164) [is](https://github.com/material-components/material-web/pull/5165) [a](https://github.com/material-components/material-web/pull/5168) [bug](https://github.com/material-components/material-web/pull/5169) [I'm](https://github.com/material-components/material-web/pull/5170) [currently](https://github.com/material-components/material-web/pull/5173) [fixing](https://github.com/material-components/material-web/pull/5177)! 😅

By default the button should truncate the label with ellipsis (...). Once I'm finished, folks will be able to make button labels wrap by adding `text-wrap: wrap` to the button in CSS.

**Reply by asyncliz** [reply]
This is actually surprisingly complicated, and has needed a lot of refactor changes to support 🙃

**Reply by treeder** [reply]
Ok, good to know, will keep an eye on it! Thanks.

**Reply by tve** [reply]
has this fix been released?

**Reply by asyncliz** [reply]
We're planning a release this week, but you can try it now with `npm install @material/web@nightly`

**Reply by treeder** [reply]
Seems this still isn't fixed in the latest release.

**Reply by asyncliz** [reply]
@treeder it's working for me: https://jsbin.com/dujuguyubi/edit?html,output

Do you have a repro?

**Reply by treeder** [reply]
@asyncLiz yes, just make any button or menu that has a space in it, eg: 

![image](https://github.com/material-components/material-web/assets/75826/2f4977cd-a878-4d86-94ff-11ea2bd94e5f)

```html
<md-menu id="user-menu" 
                  anchor="anchorbutton"
                  ?open=${this.open}
                  @closed=${() => (this.open = false)}
                  >
                  <md-menu-item href="/orders">
                    <div slot="headline">Orders</div>
                    <md-icon class="startIcon" slot="start">orders</md-icon>
                  </md-menu-item>

                    <md-menu-item id="goto-orgs" href="/orgs">
                        <div slot="headline">My Stores</div>
                        <md-icon class="startIcon" slot="start">storefront</md-icon>
                    </md-menu-item>
                    <md-menu-item id="goto-signout" @click=${this.signOut}>
                        <div slot="headline">Sign out</div>
                        <md-icon class="startIcon" slot="start">logout</md-icon>
                    </md-menu-item> 
                </md-menu>
```

**Comment by datvm** [comment]
Not sure if this is the same but I noticed `<md-*-button>` are very prone to this in a horizontal flex box and I usually have to add `flex-shrink: 0` to them to prevent them wrapping.

**Reply by asyncliz** [reply]
That should mostly be fixed now at head. Last part to this is getting the button's height to automatically grow when it wraps.

**Reply by datvm** [reply]
That's awesome! I thought the button height must be fixed due to the specs/guideline (and their CSS is fixed at 40px I think) so I had to try to keep the texts very short for mobile UIs. If they can grow that's great news!

**Reply by asyncliz** [reply]
By default they'll be single line truncated, and design still heaaavily discourages multi-line buttons, so it's an opt-in.

Buuut nature of the web, screen sizes, and localization means its a feature many folks need.
